ledger/participant-state/kvutils/src/main/protobuf/daml_kvutils.proto

```diff
@@ -0,0 +1,165 @@
+// Might want to use a more telling prefix.
+
+syntax = "proto3";
+package com.daml.ledger.participant.state.kvutils;
+option java_package = "com.daml.ledger.participant.state.kvutils";
+
+import "google/protobuf/empty.proto";
+import "google/protobuf/timestamp.proto";
+import "google/protobuf/duration.proto";
+import "da/daml_lf.proto";
+import "daml-lf/transaction/src/main/protobuf/transaction.proto";
+
+// The ledger implementation stores the `DamlLogEntry`s in an
+// ordered sequence of keys, allowing for offset-based querying.
+//
+// Each entry can be converted into a participant state `Update` event.
+message DamlLogEntry {
+  // The record time at which this entry was committed.
+  google.protobuf.Timestamp record_time = 1;
+
+  oneof payload {
+    // A committed DAML transaction.
+    // Can be converted into `Update.TransactionAccepted`.
+    DamlTransactionEntry transaction_entry = 2;
+
+    // A rejected command.
+    DamlRejectionEntry rejection_entry = 3;
+
+    // A DAML-LF archive.
+    // Can be converted into `Update.PublicPackageUploaded`.
+    daml_lf.Archive archive = 4;
+
+    // A DAML configuration change.
+    DamlConfigurationEntry configuration_entry = 5;
+  }
+}
+
+message DamlLogEntryId {
+  // Entry identifiers are opaque bytestrings chosen by the implementation that provide a unique
+  // reference to the entry.
+  // The entry identifier is used:
+  // * to lookup the log entry when producing `Update`.
+  // * as the DAML transaction identifier and rendered as hexadecimal.
+  // * as part of absolute contract identifiers.
+  bytes entry_id = 1;
+}
+
+
+// A submission to the ledger: a payload and its inputs if any.
+message DamlSubmission {
+  repeated DamlLogEntryId input_log_entries = 1;
+  repeated DamlStateKey input_daml_state = 2;
+  oneof payload {
+    DamlTransactionEntry transaction_entry = 3;
+    daml_lf.Archive archive = 4;
+    DamlConfigurationEntry configuration_entry = 5;
+  }
+}
+
+message DamlSubmitterInfo {
+  string submitter = 1;
+  string command_id = 2;
+  string application_id = 3;
+  google.protobuf.Timestamp maximum_record_time = 4;
+}
+
+message DamlTransactionEntry {
+  // The original submitted transaction, with relative identifiers.
+  // For now we're deduplicating the data and storing the contract instances
+  // both in this transaction, and in "DamlState". Later on the transaction here
+  // will only record the node identifiers, and nodes themselves are stored under
+  // "DamlState".
+  com.digitalasset.daml.lf.transaction.Transaction transaction = 1;
+
+  DamlSubmitterInfo submitter_info = 2;
+
+  // Transaction metadata
+  string workflow_id = 3;
+  google.protobuf.Timestamp ledger_effective_time = 4;
+}
+
+message DamlRejectionEntry {
+  DamlSubmitterInfo submitter_info = 1;
+
+  oneof reason {
+    // NOTE(JM): Not all payloads are used, but marked as 'string' for
+    // extensibility.
```

Comment:
How do you mean that?

Reply:
We did a pretty rough pass with the rejection reasons and I find it more than likely that we'll add detailed messages to some of them. If I mark the reasons here with google.protobuf.Empty I'll need to deprecate the field and introduce new one if we want to add a detailed message later. This feels messier to me.

We could of course just change rejection reason to always carry an extra detail string?